Mh widget

packiaVaradhan opened the pull request:
## Proposed Changes

Reason 
The widget will be embedded to a website.  The widget itself contains From and To fields, and after clicking search button , a new window should open up to the web route planner having those To and From already filled out.

Asana task https://app.asana.com/0/1111242349563886/1182118540145652

How to intregrate:

Widget.css and widget.js are necessary files for this project. Those files will be deployed on our server.  The widget.css and widget.js file's url should be given to the customer. The customer include those 2 files url in the customer website.

Url’s:
http(s)://<host:port>/js/widget.js
http(s)://<host:port>/css/widget.css


Here is the sample html client page which I used for my local test. In real time localhost:8080 should be pointing to our production or dev server host.

```<html>
<head>
<link
rel="stylesheet"
href="http://localhost:8080/css/widget.css"
/>
</head>
<body class="container">
<div>
</div>
<div id="reittiopas-mh-widget"></div>
<script src="http://localhost:8080/js/widget.js"></script>
</body>
</html>
```

## Pull Request Check List

  - A reasonable set of unit tests is included
  - Console does not show new warnings/errors
  - Changes are documented or they are self explanatory
  - This pull request does not have any merge conflicts
  - All existing tests pass in CI build
  - Code coverage does not decrease (unless measured incorrectly)

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

olavim committed:
chore: webpack config for shareable widget js

packiaVaradhan committed:
feat(mh-widget): added standalone widget

packiaVaradhan committed:
feat(mh-widget): added widget autosuggest component

packiaVaradhan committed:
feat(mh-widget): dotted border modified for input boxes

packiaVaradhan committed:
feat(mh-widget): dotted border modified for input boxes

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -869,6 +870,151 @@ export const executeSearch = (getStore, refPoint, data, callback) => {
+  let endpointSearchesPromise;
+  let searchSearchesPromise;
+  // const endpointLayers = ["CurrentPosition", "FavouritePlace", "FavouriteStop", "OldSearch", "Geocoding", "Stops"]; //layers || getAllEndpointLayers();
+  const endpointLayers = ['Geocoding'];
```
What is the idea with these `endpointLayers`? There's a whole bunch of conditionals based on them but they are defined as a constant that never changes here? 😕

packiaVaradhan replied:
i removed it .

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -869,6 +870,151 @@ export const executeSearch = (getStore, refPoint, data, callback) => {
+  const searchSearches = { type: 'search', term: input, results: [] };
+
+  let endpointSearchesPromise;
+  let searchSearchesPromise;
```
Nothing is ever set into `searchSearchesPromise`, can it be removed completely?

Tapppi commented on app/widget.js:
```diff
@@ -0,0 +1,112 @@
+        <div dangerouslySetInnerHTML={this.createSpriteMarkup()} />
+        <div className="widget-body">
+          <div className="autosuggest-panel">
+            <div className="main-title">Majoitushaku</div>
```
"Majoitushaku" literally translated means "Accommodation search". Was this supposed to say "Matkahaku" ("Trip search") or something❓

Tapppi commented on package.json:
```diff
@@ -74,7 +74,7 @@
   "lint-staged": {
     "*.{js,jsx}": [
-      "eslint"
+      "eslint --fix"
```
I don' think this will work without also adding `git add` after it, because it modifies files. Is there a need to do this `--fix` here or can you run `eslint --fix` manually when needed?

Tapppi reviewed (changes requested):
Works fine when tested locally, great work 🎉 . I went through it as best as I could, but I am quite unfamiliar with this kind of front-end code. The title of the widget is wrong though, I'm pretty sure of that one thing 😄  After Olavi/Tommi are back from holiday, I think someone else should briefly look over this as well.

Tapppi commented:
I also got this in my terminal when running the main webpage and playing around with widget, but I don't think this is related to this widget, since it is happening in terminal with node process and not on the webpage the widget is on 🤔 :
```
RelayPendingQueryTracker: Expected response to be an object, got `null`.
RelayPendingQueryTracker: Expected response to be an object, got `null`.
{ Invariant Violation: RelayPendingQueryTracker: Expected response to be an object, got `null`.
    at invariant (/opt/digitransit-ui/node_modules/fbjs/lib/invariant.js:42:15)
    at /opt/digitransit-ui/node_modules/react-relay/lib/RelayPendingQueryTracker.js:124:36
    at /opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:107:25
    at /opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:113:14
    at RelayTaskQueue._invokeWithinScopedQueue (/opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:153:7)
    at /opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:143:14
    at RelayTaskQueue._scheduleIfNecessary (/opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:170:9)
    at RelayTaskQueue.enqueue (/opt/digitransit-ui/node_modules/react-relay/lib/RelayTaskQueue.js:119:10)
    at PendingFetch._handleQuerySuccess (/opt/digitransit-ui/node_modules/react-relay/lib/RelayPendingQueryTracker.js:122:68)
    at tryCallOne (/opt/digitransit-ui/node_modules/promise/lib/core.js:37:12) name: 'Invariant Violation', framesToPop: 1 }
```

packiaVaradhan committed:
fix(mh-widget): code cleanup in searchutils, name change in search form

packiaVaradhan committed:
fix(mh-widget): code cleanup in searchutils

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -869,6 +814,68 @@ export const executeSearch = (getStore, refPoint, data, callback) => {
+    }
+    if (searchSearches && Array.isArray(searchSearches.results)) {
+      results.push(...searchSearches.results);
+    }
```
`searchSearches` is only used here, i.e. it will never be filled with any results, so the variable can be completely removed?

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -16,32 +16,6 @@ import mapPeliasModality from './pelias-to-modality-mapper';
 import { getConfiguration } from '../config';
 
-function useKyytiGeocoding(config) {
-  return (
-    !config.URL.GEOCODING_BASE_URL.includes('digitransit') &&
-    !config.URL.GEOCODING_BASE_URL.includes('mock_api_url')
-  );
-}
-
-function mapKyytiAddrToFeature(res) {
-  return {
-    type: 'Feature',
-    geometry: {
-      type: 'Point',
-      coordinates: [res.location.lon, res.location.lat],
-    },
-    properties: {
-      gid: res.gid,
-      layer: 'locality',
-      name: res.name,
-      country: res.country,
-      country_a: res.country,
-      locality: res.city,
-      label: res.title,
-    },
-  };
-}
```
All the Kyyti geocoding stuff in this file shouldn't be deleted? At least for me now the geocoding on the main route planner webpage does not work. The widget stuff shouldn't be working, but does and looks to be connecting to digitransit api instead of our own geocoding API. I looked at this in the network tab of the browser developer console to confirm. That needs to be fixed, so that we use our own geocoding api in the widget too.

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -338,42 +312,12 @@ export function getGeocodingResult(
   }
 
-  if (useKyytiGeocoding(config)) {
-    const query = {
-      text: opts.text,
-    };
-
-    if (opts['boundary.rect.min_lon']) {
-      query.bbox = `${opts['boundary.rect.min_lon']},${
-        opts['boundary.rect.min_lat']
-      },${opts['boundary.rect.max_lon']},${opts['boundary.rect.max_lat']}`;
-    }
-    if (!query.bbox) {
-      query.bbox = '19,59,32,71';
-    }
-
-    return getJson(`${config.URL.GEOCODING_BASE_URL}/places/v2/search`, query, {
-      'Kyyti-App-AppId': config.appBundleId,
-      'Accept-Language': opts.lang,
-    }).then(res => {
-      const response = {
-        features: res.search.map(mapKyytiAddrToFeature),
-      };
-      return mapPeliasModality(response.features, config);
-    });
-  }
```
This kyyti geocoding stuff is needed, ref. [other comment](https://github.com/kyyticom/digitransit-ui/pull/6#discussion_r461539872) on geocoding?

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -338,42 +312,12 @@ export function getGeocodingResult(
 
 export function searchPlace(ids, config) {
-  if (useKyytiGeocoding(config)) {
-    return Promise.resolve({});
-    // return getJson(`${config.URL.GEOCODING_BASE_URL}/places/v2/nearby/${ids}`);
-  }
```
This kyyti geocoding stuff is needed, ref. [other comment](https://github.com/kyyticom/digitransit-ui/pull/6#discussion_r461539872) on geocoding?

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -870,128 +814,45 @@ export const executeSearch = (getStore, refPoint, data, callback) => {
-          return modifiedResultsArray;
+  const language = 'fi';
+  const searchComponents = [];
+  const focusPoint =
+    config.autoSuggest.locationAware && position.hasLocation
+      ? {
+          'focus.point.lat': position.lat.toFixed(2),
+          'focus.point.lon': position.lon.toFixed(2),
```
This if clause will never evaluate to `true` because it looks at `position.hasLocation` and position [is declared](https://github.com/kyyticom/digitransit-ui/pull/6/files#diff-6200a9cfe751d20be96e10898bb53f76R818) above it as an empty object without such flag. As far as I understand this is used in widget so doesn't really matter and could be removed completely? Or what is the intended functionality?

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -1032,22 +894,5 @@ export const withCurrentTime = (getStore, location) => {
-        features: [mapKyytiAddrToFeature(res)],
-      };
-    });
-  }
```
This kyyti geocoding stuff is needed, ref. [other comment](https://github.com/kyyticom/digitransit-ui/pull/6#discussion_r461539872) t on geocoding?

Tapppi commented on app/widget.scss:
```diff
@@ -9361,12 +9361,9 @@ input.react-autosuggest__input {
     box-shadow: none;
-    /* First apply outline that is used in Edge and Firefox */
     outline: 5px auto -webkit-focus-ring-color;
-    /* Then override outline for webkit browsers like Chrome, this has no effect on other browsers */
```
Is there a reason to delete these comments? They seem helpful 😄

Tapppi reviewed (changes requested):
Looking better, but the Kyyti geocoding surely shouldn't have been deleted 🤔

packiaVaradhan committed:
fix(mh-widget): kyyti geocoding stuff addded back and code cleanup

packiaVaradhan committed:
fix(mh-widget): kyyti geocoding stuff addded back in searchutils

packiaVaradhan committed:
fix(mh-widget): kyyti geocoding stuff addded back in searchutils

Tapppi commented on app/util/searchUtils.js:
```diff
@@ -869,6 +869,57 @@ export const executeSearch = (getStore, refPoint, data, callback) => {
 };
 
+export function executeWidgetSearchImmediate({ input }, callback) {
+  const config = getConfiguration(null);
```
The widget is still using digitransit API For place searches as can be seen in the request url of network tab in below screenshot. 

![image](https://user-images.githubusercontent.com/3275280/88731243-b64f0c00-d13f-11ea-8376-1834d6014967.png)

It seems to be caused by default config being loaded here, because `null` is equivalent to `'default'` in the `getConfiguration()` function. The `executeSearchImmediate` function above [seems to get config as param](https://github.com/kyyticom/digitransit-ui/blob/0944cd42fa4b8b3a90d14828d0d7aaffe7fe539f/app/util/searchUtils.js#L688) which is setup in the React component in [some kind of context](https://github.com/kyyticom/digitransit-ui/blob/0944cd42fa4b8b3a90d14828d0d7aaffe7fe539f/app/component/DTSearchAutosuggest.js#L173). It looks like the `getConfiguration` function is intended to be used in `server/` components and not `app/` components.

Unfortunately my expertise with React stops here and I have no idea how we get the correct configuration. Can we set the correct config somehow in the actual server code that serves `widget.js` or some kind of variable in the widget embed HTML 🤔 Probably same way can be used as setting config for the actual reittiopas webpage..

packiaVaradhan replied:
yes its using default config.  context can't be used in widget.
What configuration file should be used for this widget?  should it be "mhdev"(config.mhdev.js) configuration? if yes,it should be injected to the widget during build. So that, the widget.js component will be having correct configurations..

Tapppi reviewed (changes requested):
Everything else is looking good to me, but the widget is still not using the correct config so it ends up calling DigiTransit place search instead of our own place search. That needs to be fixed somehow, hopefully your better knowledge of React helps there.

packiaVaradhan committed:
fix(mh-widget): added config file for mh-widget under configurations

Tapppi committed:
refactor(widget): inject app config into widget during runtime

Instead of injecting configuration during build and using separate
config, inject the configuration during runtime when the file is
fetched. This allows generic builds, but does leak the whole config
codebase to the client..

Tapppi commented:
@packiaVaradhan Hey, after testing your changes I ended up using the app config for widget as well. The build `definePlugin` would've been good otherwise, but we don't have config specific build phase, so would've required new build configuration. Instead I asked a little help from Olavi to get the config in there during runtime. Please check and test the changes, but works for me locally now! 🤞 😊

packiaVaradhan commented:
> @packiaVaradhan Hey, after testing your changes I ended up using the app config for widget as well. The build `definePlugin` would've been good otherwise, but we don't have config specific build phase, so would've required new build configuration. Instead I asked a little help from Olavi to get the config in there during runtime. Please check and test the changes, but works for me locally now! 🤞 😊

for me it works locally too . i am merging it now.
